bank/persistence/mongo.go

```diff
@@ -0,0 +1,287 @@
+		return err
+	}
+
+	return nil
+}
+
+func (se *MongoSessionStore) DeleteAccount(userID string, accountID string) error {
+	session := se.Session.Clone()
+	defer session.Close()
+
+	var result domain.Account
+
+	err := session.DB(se.DBname).C("accounts").Find(bson.M{"userid": userID, "accountid": accountID}).One(&result)
+	if err != nil {
+		return err
+	}
+
+	err = session.DB(se.DBname).C("accounts").Remove(bson.M{"userid": userID, "accountid": accountID})
+	if err != nil {
+		return err
+	}
+
+	err = session.DB(se.DBname).C("history").Insert(&domain.History{AccountID: accountID, UserID: userID, TransactionType: "account deleted", Currency: result.Currency, Date: time.Now().Local()})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (se *MongoSessionStore) StartSession(u domain.User, userSessionDuration time.Time) (*domain.Session, error) {
+	session := se.Session.Clone()
+	defer session.Close()
+
+	var result domain.Session
+
+	err := session.DB(se.DBname).C("sessions").Insert(&domain.Session{Expires: userSessionDuration, SessionID: RandString(30), UserID: u.UserID})
+	if err != nil {
+		return nil, err
+	}
+	err = session.DB(se.DBname).C("sessions").Find(bson.M{"userid": u.UserID}).One(&result)
+	if err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+func (se *MongoSessionStore) FindSessionAvailableAt(sessionID string, instant time.Time) (*domain.Session, bool) {
+	session := se.Session.Clone()
+	defer session.Close()
+
+	var result domain.Session
+
+	session.DB(se.DBname).C("sessions").Find(bson.M{"sessionid": sessionID}).One(&result)
+
+	if result.UserID != "" && result.SessionID == sessionID && !instant.After(result.Expires) {
+		return &result, true
+	}
+
+	return &result, false
+}
+
+func (se *MongoSessionStore) DeleteSession(sessionID string) error {
+	session := se.Session.Clone()
+	defer session.Close()
+
+	err := session.DB(se.DBname).C("sessions").Remove(bson.M{"sessionid": sessionID})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (se *MongoSessionStore) UpdateSession(sessionID string, userSessionDuration time.Time) error {
+	session := se.Session.Clone()
+	defer session.Close()
+
+	err := session.DB(se.DBname).C("sessions").Update(bson.M{"sessionid": sessionID}, bson.M{"$set": bson.M{"expires": userSessionDuration}})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (se *MongoSessionStore) RegisterUser(req domain.UserRegistrationRequest) (*domain.User, error) {
+	session := se.Session.Clone()
+	defer session.Close()
+	var err error
+
+	user := domain.User{Username: req.Username, UserID: RandString(30), Name: req.Name, Email: req.Email, Age: req.Age}
+
+	user.Hash, err = bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)
+	if err != nil {
+		return nil, err
+	}
+
+	err = session.DB(se.DBname).C("users").Insert(&domain.User{Username: user.Username, UserID: user.UserID, Name: user.Name, Email: user.Email, Age: user.Age, Hash: user.Hash})
+	if err != nil {
```

Comment:
Make a query instead of verifying of the constraint error of mongodb. 

Something like: 
session.DB(se.DBName)C("users").Query(bson.M{"username", user.Username}).One(&user)

and if record is returned then you can return domain.ErrUserAlreadyExists.